raidboss: r4s strategy options for NA strats

jacob-keller opened the pull request:
This series includes several strategy options for the m4s fight that fit the
standard NA PF strategies.

- **raidboss: add support for DN strategy for Ion Cluster in m4s**
- **raidboss: add support for DN strategy during Witch Hunt in m4s**
- **raidboss: add uptime sunrise sabaath callouts in m4s**

These are based off the shabin pastebin and hector guide, with the uptime
waymarks strategy for sunrise sabaath.

I made all of these configuration options, so that existing default callouts
which do not use a specific strategy are not affected. I tried to put good
brief explanations, and used the standard strat names I see going around.

I've tested variations of this code in my runs, though it did not use a
strategy option at the time. Thus, this exact code is not yet battle tested
outside of the simulator.

xiashtra commented:
The Sunrise callouts could be modified to also call wall baits as an option too; the tower/cannon priority is the same and only the baiting direction is different.

wexxlee commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -275,6 +382,60 @@ export interface Data extends RaidbossData {
+      options: {
+        en: {
+          'None - Just call debuff and both tower spawns.': 'none',
+          'Uptime - Call specific cannon and tower spots using priority system (DPS N and CW, Support NW and CCW). Assumes AutoCAD uptime waymarks.':
```
Not going to be able to look at this for a while, but on a quick look, this caught my eye -- the prop key is what goes in the select box, and this is a long string to have inside a dropdown.  For this and the others, I'd suggest much shorter names, and then use the additional property (I think it's `comment`?) to add an explanatory comment and link to the strats -- see p12s for examples.

jacob-keller replied:
Sure. I wasn't quite sure how to explain, since I wanted to include enough information for folks to find and ensure they are using the right strategy. These are somewhat specific assumptions about order and spots.

wexxlee replied:
Oh, yeah I think explanations are super helpful - definitely not suggesting to remove them or anything.  I think `comment` was added for just that reason so we could explain/link to strat options & explanations.

jacob-keller commented:
> The Sunrise callouts could be modified to also call wall baits as an option too; the tower/cannon priority is the same and only the baiting direction is different.

Ya. I can probably include that as another option. My group (and most PFs) seem to do uptime strategy so that's what I implemented.

xiashtra commented:
> > The Sunrise callouts could be modified to also call wall baits as an option too; the tower/cannon priority is the same and only the baiting direction is different.
> 
> Ya. I can probably include that as another option. My group (and most PFs) seem to do uptime strategy so that's what I implemented.

I see a good mix of both in PF which is why I suggested it.

jacob-keller commented:
> > > The Sunrise callouts could be modified to also call wall baits as an option too; the tower/cannon priority is the same and only the baiting direction is different.
> > 
> > 
> > Ya. I can probably include that as another option. My group (and most PFs) seem to do uptime strategy so that's what I implemented.
> 
> I see a good mix of both in PF which is why I suggested it.

I've updated the descriptions, renamed the strategy as "snakes prio" to match the way its discussed in the various sites. I split the uptime callouts from the priority system so now it will call the direction to bait properly for non-uptime spots.

Hopefully this version looks good! I'll be testing it in a live run tonight.

wexxlee commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -227,7 +246,96 @@ const swordQuiverOutputStrings = {
+  proximateCurrent: {
+    en: 'Front of Middle (Near Cone)',
+  },
+  spinningConductorSupport: {
+    en: 'Left Rivet (Small AoE)',
+  },
+  spinningConductorDPS: {
+    en: 'Right Rivet (Small AoE)',
+  },
+  roundhouseConductorSupport: {
+    en: 'Left Rivet (Donut AoE)',
+  },
+  roundhouseConductorDPS: {
+    en: 'Right Rivet (Donut AoE)',
+  },
+  colliderConductor: {
+    en: 'Middle Rivet (Get Hit by Cone)',
```
Sorta wondering if 'rivet' is really necessary? It seems like left or right would communicate the necessary information (or perhaps even 'Front Left' or 'Front Right'.  I think 'Middle Rivet' may also be confusing since it doesn't clearly communicate position relative to the `current` debuff -- maybe 'Back Middle' so it's more parallel with the `current` callouts?

jacob-keller replied:
Supposedly the position is fairly precise on the rivets, which is why I used it.

jacob-keller replied:
In specific, the player in the middle must be nearest to the cone player.

jacob-keller replied:
What about:

Front Left,
Front Right,
Front of Middle
Middle, Behind Current

wexxlee commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -275,6 +383,84 @@ export interface Data extends RaidbossData {
+      comment: {
+        en: `Strategy for resolving Sunrise Sabbath.
+
+             None - Call debuff, both tower spawns, and both matching towers.,
+             Snake Prio - Use the snakes priority system with support looking for
+                          a matching tower or cannon starting at North West and going
+                          counter clockwise. DPS look for matching tower or cannon
+                          starting North and going clockwise.`,
```
same

wexxlee commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -275,6 +383,84 @@ export interface Data extends RaidbossData {
+        en: 'Witch Hunt Bait Strategy',
+      },
+      comment: {
+        en: `Strategy for baiting Witch Hunt AoEs.
+
+             None - Call both party and bait positions, no strategy.
+             DN - Flexible priority to maximize melee uptime, with tanks taking
+                  first near bait, healers taking first far bait, melees taking
+                  second near bait, and ranged taking second far bait.`,
```
I'm not sure if the whitespace will get properly translated/formatted in HTML.  Is there an existing trigger where we do something similar that we can check to confirm proper UI display?

jacob-keller replied:
It does not, I checked. I'll see if I can come up with a better thing. Perhaps I can use html breaks to separate each strategy?

wexxlee commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -1931,19 +2186,65 @@ const triggerSet: TriggerSet<Data> = {
           const cannonLocs = task === 'yellowShort' ? blueCannons : yellowCannons;
+
+          if (data.triggerSetConfig.sunrise === 'snakePrio') {
+            const dpsPrio: DirectionOutputIntercard[] = ['dirNE', 'dirSE', 'dirSW', 'dirNW'];
+            const supPrio: DirectionOutputIntercard[] = ['dirNW', 'dirSW', 'dirSE', 'dirNE'];
```
```suggestion
            const dpsPrio: DirectionOutputIntercard[] = ['dirNE', 'dirSE', 'dirSW'];
            const supPrio: DirectionOutputIntercard[] = ['dirNW', 'dirSW', 'dirSE'];
```

With snake prio, I don't think the fourth spot will ever be used, right?

jacob-keller replied:
Nope. I can remove it.

wexxlee commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -2001,6 +2302,21 @@ const triggerSet: TriggerSet<Data> = {
+        westTower: {
+          en: 'West Tower',
+        },
+        specificCannon: {
+          en: '${loc} Cannon - Bait ${bait}',
+        },
```
Wouldn't it be possible to use `yellowShort` and `blueShort` for this?

jacob-keller replied:
i'll see. I originally had a different text string for both strategies, and I didn't find the tower color callout to be helpful but perhaps others might like it since they can confirm they have the right color of tower.

jacob-keller replied:
And it would make the code simpler.

wexxlee reviewed (commented):
Overall looks promising.  My group is done with reclears, so not sure when I will find time to get back in to test -- but a few initial comments in the meantime.

jacob-keller commented:
![image](https://github.com/user-attachments/assets/cd806999-8a4b-4d5d-8a5d-4b7cd824e823)
Here's example callouts for witch hunt.

jacob-keller commented:
![image](https://github.com/user-attachments/assets/61fe4688-5e5a-43ee-a946-e15e79f68a10)

![image](https://github.com/user-attachments/assets/679ab4d0-345d-4b1a-ba95-8633a261c34d)

![image](https://github.com/user-attachments/assets/90e09a9b-61eb-47c0-b9e9-8a1c7dafde3e)

![image](https://github.com/user-attachments/assets/4d967290-72e1-4b3e-a6b1-920f38cfc737)

![image](https://github.com/user-attachments/assets/21cbb917-0750-4510-a933-1ac60cceed7b)

![image](https://github.com/user-attachments/assets/4dd7e529-fb1f-4eaa-95cf-789ea769dea6)

![image](https://github.com/user-attachments/assets/cea967e8-1edf-45b8-8949-942c83d32eb8)

![image](https://github.com/user-attachments/assets/f85fb10f-7acf-4c5d-875b-8574968ccdce)

Here's the current example of callouts for sunrise sabbath

jacob-keller commented:
![image](https://github.com/user-attachments/assets/bf97f123-70dc-482f-936a-f1ecbc3a4e88)

Here's the config text. It's definitely ugly. Let me try to fix that.

jacob-keller commented:
![image](https://github.com/user-attachments/assets/9daea992-9e59-4cee-8c93-2419bf30e93d)

Here's the new version with HTML breaks!

jacob-keller commented:
![image](https://github.com/user-attachments/assets/ca4d60fe-5612-476e-b714-f024c6621430)

This is if I re-use yellowShort and and blueShort.. I don't like "Point W"

xiashtra commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -537,11 +720,21 @@ const triggerSet: TriggerSet<Data> = {
+          else
+            baitOrder = ['near', 'far', 'near', 'far'];
+        } else if (data.witchHuntBait === 'far') {
+          if (data.triggerSetConfig.witchHunt === 'DN')
+            baitOrder = ['far', 'near', 'far', 'near'];
+          else
+            baitOrder = ['healer', 'tank', 'ranged', 'melee'];
```
```suggestion
          if (data.triggerSetConfig.witchHunt === 'DN')
            baitOrder = ['healer', 'tank', 'ranged', 'melee'];
          else
            baitOrder = ['far', 'near', 'far', 'near'];
```

jacob-keller replied:
Oh good call!

jacob-keller replied:
This should be fixed now thanks for spotting it!

jacob-keller commented:
![image](https://github.com/user-attachments/assets/41cdf37c-5719-45ff-ada6-04a063f376df)

Here's what I came up with. I think this is acceptable

xiashtra commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -275,6 +383,81 @@ export interface Data extends RaidbossData {
+      comment: {
+        en: `Strategy for resolving debuffs during Ion Cluster.
+
+             None - Call the debuff only, no strategy.
+             DN - use rivet positions based on the shabin pastebin.`,
```
```suggestion
             None: Call the debuff only, no strategy.
             DN: Use rivet positions based on the shabin pastebin.`,
```
Format consistency.

jacob-keller replied:
Fixed.

xiashtra commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -275,6 +383,81 @@ export interface Data extends RaidbossData {
+      comment: {
+        en: `Strategy for resolving Sunrise Sabbath.<br>
+             None: Call debuffs, both tower spawns, and matching towers.<br>
+             Snakes Prio - Popular priority system used in NA PF. Support players
```
```suggestion
             Snakes Prio: Popular priority system used in NA PF. Support players
```
Format consistency.

jacob-keller replied:
Fixed.

jacob-keller commented on ui/raidboss/data/07-dt/raid/r4s.ts:
```diff
@@ -153,6 +154,24 @@ const witchHuntAlertOutputStrings = {
     cn: '引导远 (小队近)',
     ko: '멀리 유도 (본대 가까이)',
   },
+  baitNear: {
+    en: 'Close (Party Far)',
```
I had originally intended to call "Close" for all of the options, but that doesn't work because of melee/ranged issues due to parties not being perfect composition. Makes sense in that case to just always call the role. ok.

jacob-keller commented:
@xiashtra can you re-check this? I think I've fixed all your concerns. I also took the time to squash everything down since I had a lot of fixes that made reviewing individual commits more difficult.

jacob-keller commented:
I updated the witchhunt callout to always put bait first. I think that's easier to parse, and certainly easier for a shot caller to use.

jacob-keller commented:
I'd like to know if anyone thinks this needs more changes :)

xiashtra reviewed (approved):
I haven't been able to test this as much as I would have liked (groups are regularly skipping Sunrise), but everything seems to be working ok from what limited testing I was able to do.
